### PR TITLE
 [STRY/BE] REST API: Daily Journal Entries

### PR DESCRIPTION
**Pull request by zhl024**
Second try to merge

### CODE REVIEW

**Review by dino-inc — APPROVED** [review]
The first pull request did not merge properly, so this is try # 2.

File: source/routes/update/daily.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
```
Comment (b2wu) [inline source/routes/update/daily.js:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const db = new PouchDB("db");
```

File: source/routes/create/daily.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+*/
+module.exports = {
+    '/create/daily': {
+        methods: ['post'],
+        fn: function(req, res, next) {
+            // Check if every field exists, if not, throw error
+            let requiredFields = ["user", "date", "docType", "monthKey", "bullets"];
+            requiredFields.forEach((jsonField, index)=>{
+                if(!req.body[jsonField]){
+                    throw new Error('MISSING FIELD');
```
Comment (b2wu) [inline source/routes/create/daily.js:25]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
                    throw new Error("MISSING FIELD");
```

File: source/routes/update/daily.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+const PouchDB = require('pouchdb');
```
Comment (b2wu) [inline source/routes/update/daily.js:1]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const PouchDB = require("pouchdb");
```

File: source/routes/read/daily.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    "date": "2021-05-09",
+    "monthKey": "05-09",
+    "bullets": [
+        {bullet1 json objectt},
+        {bullet2 json object}
+    ]
+}
+*/
+module.exports = {
+    '/read/daily': {
```
Comment (b2wu) [inline source/routes/read/daily.js:17]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
    "/read/daily": {
```

File: source/routes/delete/daily.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
+
+/* *
+Delete daily entry json object and all bullet documents in that daily entry.
+*/
+module.exports = {
+    '/delete/daily': {
+        methods: ['delete'],
```
Comment (b2wu) [inline source/routes/delete/daily.js:10]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        methods: ["delete"],
```

File: source/routes/read/daily.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
```
Comment (b2wu) [inline source/routes/read/daily.js:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const db = new PouchDB("db");
```

File: source/routes/create/daily.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    "user": "dave",
+    "date": "2021-05-09",
+    "docType": "dailyJournal",
+    "monthKey": "May",
+    "bullets": ["bullet-id1", "bullet-id2", ...]
+}
+*/
+module.exports = {
+    '/create/daily': {
+        methods: ['post'],
```
Comment (b2wu) [inline source/routes/create/daily.js:19]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        methods: ["post"],
```

File: source/routes/update/daily.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ {
+    "_id": "documentID", 
+    "updateField": {
+        "customFieldToUpdate": "customDataToUpdate",
+        
+    }
+}
+*/
+module.exports = {
+    '/update/daily': {
```
Comment (b2wu) [inline source/routes/update/daily.js:17]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
    "/update/daily": {
```

File: source/routes/delete/daily.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
```
Comment (b2wu) [inline source/routes/delete/daily.js:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const db = new PouchDB("db");
```

File: source/routes/delete/daily.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
```
Comment (b2wu) [inline source/routes/delete/daily.js:2]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
PouchDB.plugin(require("pouchdb-find"));
```

File: source/routes/create/daily.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            .then((response) => {
+                res.send(response);
+            })
+            .catch((err) => {
+                console.log(err);
+                res.send("error");
+            });
+        }
+    }
+}
```
Comment (b2wu) [inline source/routes/create/daily.js:49]: Codacy has a fix for the issue: [Missing semicolon.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
}; 
```

File: source/routes/create/daily.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    "bullets": ["bullet-id1", "bullet-id2", ...]
+}
+*/
+module.exports = {
+    '/create/daily': {
+        methods: ['post'],
+        fn: function(req, res, next) {
+            // Check if every field exists, if not, throw error
+            let requiredFields = ["user", "date", "docType", "monthKey", "bullets"];
+            requiredFields.forEach((jsonField, index)=>{
```
Comment (b2wu) [inline source/routes/create/daily.js:23]: Codacy has a fix for the issue: [Missing space before =>.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
            requiredFields.forEach((jsonField, index) =>{
```

File: source/routes/create/daily.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
```
Comment (b2wu) [inline source/routes/create/daily.js:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const db = new PouchDB("db");
```

File: source/routes/read/daily.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+}
+*/
+module.exports = {
+    '/read/daily': {
+        methods: ['get'],
+        fn: function (req, res, next) {
+            let tempArr = [];
+            //get Daily entry documnet by id
+            db.get(req.body._id)
+            .then(response => {
```
Comment (b2wu) [inline source/routes/read/daily.js:23]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
            .then((response) => {
```

File: source/routes/update/daily.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
+const PouchDB = require('pouchdb');
```
Comment (b2wu) [inline source/routes/update/daily.js:1]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const PouchDB = require("pouchdb");
```

File: source/routes/read/daily.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
+const PouchDB = require('pouchdb');
```
Comment (b2wu) [inline source/routes/read/daily.js:1]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const PouchDB = require("pouchdb");
```

File: source/routes/delete/daily.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+            // access all bullet documents' id
+            response.bullets.forEach((bullet, index) => {
+              db.get(bullet)
+                  .then(function(doc) {
+                    // delete bullet document
+                    return db.remove(doc._id, doc._rev);
+                  })
+                  .catch((err) => {
+                    console.log(err);
+                    res.send('error');
```
Comment (b2wu) [inline source/routes/delete/daily.js:24]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
                    res.send("error");
```

File: source/routes/create/daily.js [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+*/
+module.exports = {
+  '/create/daily': {
+    methods: ['post'],
+    fn: function(req, res, next) {
+      // Check if every field exists, if not, throw error
+      const requiredFields = ['user', 'date', 'docType', 'monthKey', 'bullets'];
+      requiredFields.forEach((jsonField, index) =>{
+        if (!req.body[jsonField]) {
+          throw new Error('MISSING FIELD');
```
Comment (b2wu) [inline source/routes/create/daily.js:25]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
                    throw new Error("MISSING FIELD");
```

File: source/routes/read/daily.js [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    "date": "2021-05-09",
+    "monthKey": "05-09",
+    "bullets": [
+        {bullet1 json objectt},
+        {bullet2 json object}
+    ]
+}
+*/
+module.exports = {
+  '/read/daily': {
```
Comment (b2wu) [inline source/routes/read/daily.js:17]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
    "/read/daily": {
```

File: source/routes/read/daily.js [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+            response.bullets.forEach((bullet, index, array) => {
+              // get bullet document by id
+              db.get(bullet)
+                  .then((bulletResponse) => {
+                    tempArr.push(bulletResponse);
+                  })
+                  .catch((err) => {
+                    res.send('error');
+                  })
+                  .finally(()=>{
```
Comment (b2wu) [inline source/routes/read/daily.js:38]: Codacy has a fix for the issue: [Missing space before =>.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
                    .finally(() =>{
```

File: source/routes/create/daily.js [hunk 21]
```diff
@@ -0,0 +1,49 @@
+const PouchDB = require('pouchdb');
```
Comment (b2wu) [inline source/routes/create/daily.js:1]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const PouchDB = require("pouchdb");
```

File: source/routes/delete/daily.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+                  });
+            });
+            // delete daily entry
+            db.remove(response._id, response._rev);
+            res.send('success');
```
Comment (b2wu) [inline source/routes/delete/daily.js:29]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
            res.send("success");
```

File: source/routes/create/daily.js [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+PouchDB.plugin(require('pouchdb-find'));
```
Comment (b2wu) [inline source/routes/create/daily.js:2]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
PouchDB.plugin(require("pouchdb-find"));
```

File: source/routes/read/daily.js [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+    "monthKey": "05-09",
+    "bullets": [
+        {bullet1 json objectt},
+        {bullet2 json object}
+    ]
+}
+*/
+module.exports = {
+  '/read/daily': {
+    methods: ['get'],
```
Comment (b2wu) [inline source/routes/read/daily.js:18]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        methods: ["get"],
```

File: source/routes/update/daily.js [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+    "_id": "documentID",
+    "updateField": {
+        "customFieldToUpdate": "customDataToUpdate",
+
+    }
+}
+*/
+module.exports = {
+  '/update/daily': {
+    methods: ['post'],
```
Comment (b2wu) [inline source/routes/update/daily.js:18]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        methods: ["post"],
```

File: source/routes/create/daily.js (hunk 13 above)
Comment (b2wu) [inline source/routes/create/daily.js:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const db = new PouchDB("db");
```

File: source/routes/update/daily.js [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+              }
+            };
+            // Put newly updated document into the databse
+            db.put(response)
+                .then(() => {
+                  res.send('success');
+                })
+                .catch((err) => {
+                  console.log(err);
+                  res.send('error');
```
Comment (b2wu) [inline source/routes/update/daily.js:42]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
                  res.send("error");
```

File: source/routes/create/daily.js [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+    "user": "dave",
+    "date": "2021-05-09",
+    "docType": "dailyJournal",
+    "monthKey": "May",
+    "bullets": ["bullet-id1", "bullet-id2", ...]
+}
+*/
+module.exports = {
+  '/create/daily': {
+    methods: ['post'],
```
Comment (b2wu) [inline source/routes/create/daily.js:19]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        methods: ["post"],
```

File: source/routes/update/daily.js [hunk 28]
```diff
@@ -0,0 +1,51 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
```
Comment (b2wu) [inline source/routes/update/daily.js:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
const db = new PouchDB("db");
```

File: source/routes/update/daily.js [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+    }
+}
+*/
+module.exports = {
+  '/update/daily': {
+    methods: ['post'],
+    fn: function(req, res, next) {
+      // If the update request does not in form specified above, throw error
+      if (!req.body.updateField) {
+        throw new Error('MISSING UPDATE DATA');
```
Comment (b2wu) [inline source/routes/update/daily.js:22]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        throw new Error("MISSING UPDATE DATA");
```

File: source/routes/update/daily.js [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+ {
+    "_id": "documentID",
+    "updateField": {
+        "customFieldToUpdate": "customDataToUpdate",
+
+    }
+}
+*/
+module.exports = {
+  '/update/daily': {
```
Comment (b2wu) [inline source/routes/update/daily.js:17]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
    "/update/daily": {
```

File: source/routes/create/daily.js [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+    fn: function(req, res, next) {
+      // Check if every field exists, if not, throw error
+      const requiredFields = ['user', 'date', 'docType', 'monthKey', 'bullets'];
```
Comment (b2wu) [inline source/routes/create/daily.js:22]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
      const requiredFields = ["user", 'date', 'docType', 'monthKey', 'bullets'];
```

File: source/routes/delete/daily.js [hunk 32]
```diff
@@ -0,0 +1,37 @@
+const PouchDB = require('pouchdb');
+PouchDB.plugin(require('pouchdb-find'));
+const db = new PouchDB('db');
+
+/* *
+Delete daily entry json object and all bullet documents in that daily entry.
+*/
+module.exports = {
+  '/delete/daily': {
+    methods: ['delete'],
```
Comment (b2wu) [inline source/routes/delete/daily.js:10]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-sp21-group13/cse110-sp21-group13/pullRequest?prid=7433333)
```suggestion
        methods: ["delete"],
```

**Pull request merged**